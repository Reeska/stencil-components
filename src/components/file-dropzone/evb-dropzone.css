:host {
  display: flex;
  background-color: transparent;
  min-height: 2em;
  border: 2px dashed #999;
  padding: .5rem;
  transition: all .2s ease-in;
}

:host(.dropzone--active) {
  background-color: #e6e6e6;
}

:host(.dropzone--hover) {
  background-color: #f2f2f2;
}
