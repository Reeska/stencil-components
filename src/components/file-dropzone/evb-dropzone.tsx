import { Component, ComponentInterface, Element, Event, EventEmitter, Listen, Prop, State } from '@stencil/core';

import { PickedFile } from '../file-picker/pickedfile';

@Component({
  tag: 'evb-dropzone',
  styleUrl: 'evb-dropzone.css',
  shadow: true,
})
export class EvbDropzone implements ComponentInterface {

  /**
   * A string that defines the file types the file input should accept. This string is a
   * comma-separated list of unique file type specifiers.
   * To accept images, video and audio, use: accept="audio/*,video/*,image/*"
   * otherwise provide the correct mimetype, eg: image/png for png images etc
   */
  @Prop() accept: string;

  /**
   * Fired after a file has been pickedÂ§
   */
  @Event()
  dropped: EventEmitter<PickedFile>;

  /**
   * Dropzone is active when the user is dragging
   */
  @State()
  private active: boolean;

  /**
   * Indicated that the user hovers above the dropzone
   */
  @State()
  private hover: boolean;

  @Element()
  host: HTMLElement;

  private filePicker: HTMLEvbFilepickerElement;

  @Listen('window:dragstart')
  dragStart() {
    this.active = true;
  }

  @Listen('window:dragend')
  dragEnd() {
    this.active = false;
  }

  @Listen('dragover', { passive: false })
  dragOver(event: DragEvent) {
    event.preventDefault();
    event.dataTransfer.dropEffect = 'move';
  }

  @Listen('dragenter')
  dragEnter() {
    this.hover = true;
  }

  @Listen('dragleave')
  dragLeave() {
    this.hover = false;
  }

  @Listen('drop', { passive: false })
  drop(event: DragEvent) {
    event.preventDefault();
    event.stopPropagation();
    if (event.dataTransfer.files) {
      this.filePicker.handleFiles(event.dataTransfer.files);
    }

    this.active = false;
    this.hover = false;
  }

  componentDidLoad() {
    console.log('HOST', this.host);
    this.filePicker = this.host.shadowRoot.querySelector('evb-filepicker');
  }

  hostData() {
    return {
      class: {
        'dropzone--active': this.active,
        'dropzone--hover': this.hover
      },
    };
  }

  render() {
    return ([
      <evb-filepicker
        onPick={event => this.dropped.emit(event.detail)}
        accept={this.accept}></evb-filepicker>,
      <slot />
    ]);
  }
}
