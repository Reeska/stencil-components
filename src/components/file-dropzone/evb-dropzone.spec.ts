import { EvbDropzone } from './evb-dropzone';

const mockFilepicker = {
  handleFiles: jest.fn()
};

const mockHost = {
  shadowRoot: {
    querySelector: jest.fn(() => mockFilepicker)
  }
};

function createMockDragEvent(hasFile = false): DragEvent {
  const files = hasFile ? [{}] : [];
  // @ts-ignore
  return {
    stopPropagation: jest.fn(),
    preventDefault: jest.fn(),
    // @ts-ignore
    dataTransfer: { files }
  };
}

// tslint:disable: no-unbound-method
describe('EvbDropzone', () => {

  let component: EvbDropzone;

  beforeEach(() => component = new EvbDropzone());

  it('should render', () => {
    expect(component.render()).toMatchSnapshot();
  });

  it('should toggle active', () => {
    expect(isActive(component)).toBeFalsy();
    component.dragStart();
    expect(isActive(component)).toBeTruthy();
    component.dragEnd();
    expect(isActive(component)).toBeFalsy();
  });

  it('should toggle hover', () => {
    expect(isHover(component)).toBeFalsy();
    component.dragEnter();
    expect(isHover(component)).toBeTruthy();
    component.dragLeave();
    expect(isHover(component)).toBeFalsy();
  });

  it('should be able to drop a file', () => {
    // @ts-ignore just ignore error about this mock...
    component.host = mockHost;
    component.componentDidLoad();

    const mockEvent = createMockDragEvent(true);

    component.drop(mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(mockEvent.stopPropagation).toHaveBeenCalled();

    expect(mockFilepicker.handleFiles).toHaveBeenCalled();
  });

  it('should call cancel event on dragover', () => {
    // @ts-ignore just ignore error about this mock...
    const mockEvent = createMockDragEvent();

    component.dragOver(mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
  });
});

function isActive(component: EvbDropzone) {
  return component.hostData().class['dropzone--active'];
}

function isHover(component: EvbDropzone) {
  return component.hostData().class['dropzone--hover'];
}
