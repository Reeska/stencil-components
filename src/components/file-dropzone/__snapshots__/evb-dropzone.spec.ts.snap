// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbDropzone should render 1`] = `
Array [
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": Object {
      "accept": undefined,
      "onPick": [Function],
    },
    "vchildren": null,
    "vkey": undefined,
    "vname": undefined,
    "vtag": "evb-filepicker",
    "vtext": undefined,
  },
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": null,
    "vchildren": null,
    "vkey": undefined,
    "vname": undefined,
    "vtag": "slot",
    "vtext": undefined,
  },
]
`;
